README.md

```diff
@@ -1,9 +1,166 @@
+To write a new test, a developer should create a test spec file in the generator/testspecs folder, using the [format below](#test-spec-format). 
+
+They then need to write a generator function in ```generator/generator_lib.py```.  This generator function takes no arguments, and should output a tuple of python binary strings.  
+
+The first element of the tuple should be the protocol buffer encoded value of the operation message that is described in the spec.operation.body_description field of the test spec.  The second element of the tuple should be the protocol buffer encoded value of the result message that is described in the spec.result.body_description field of the test spec.
+
+An example generator for a list opcodes test is shown below:
+
+```python
+def gen_list_opcodes_auth_direct():
+    op = protobuf.list_opcodes_pb2.Operation()
+    op.provider_id = 1
+
+    result = protobuf.list_opcodes_pb2.Result()
+    result.opcodes.extend([1,3,2])
+    return (op.SerializeToString(), result.SerializeToString())
+```
+The generated protocol buffers python library files are in generator/protobuf.
+
+Finally, the developer should add an entry to the generator dictionary at the bottom of the ```generator/generator_lib.py``` file.  The key must correspond to the value in the test spec's spec.generator field.  The value in the generator dictionary must be the new generator function name.  e.g.:
+
+```python
+generators = {
+    'ping_no_auth': gen_ping_no_auth,
+    'list_opcodes_auth_direct': gen_list_opcodes_auth_direct,
+}
+```
+**NOTE**  A generator function can be used in multiple test specs, if required.
+
+## Test Spec Format
+
+Test specs are defined as YAML.  An example file is shown below:
+
+```yaml
+# Copyright 2021 Contributors to the Parsec project.
+# SPDX-License-Identifier: Apache-2.0
+spec:
+  name: list_opcodes_auth_direct
+  generator: list_opcodes_auth_direct
+  description: List opcodes using direct authentication
+  operation:
+    header:
+      magic_number: 0x5EC0A710
+      header_size: 0x1E
+      major_version_number: 0x01
+      minor_version_number: 0x00
+      flags: 0x0000
+      provider: 0x00
+      session_handle: 0x0000000000000000
+      content_type: 0x00
+      accept_type: 0x00
+      auth_type: 0x00
+      content_length: auto
+      auth_length: auto
+      opcode: 0x00000009
+      status: 0x0000
+    body_description: provider id 0x01
+    auth: 
+      type: direct
+      app_name: jimbob
+  result:
+    header:
+      magic_number: 0x5EC0A710
+      header_size: 0x1E
+      major_version_number: 0x01
+      minor_version_number: 0x00
+      flags: 0x0000
+      provider: 0x00
+      session_handle: 0x0000000000000000
+      content_type: 0x00
+      accept_type: 0x00
+      auth_type: 0x00
+      content_length: auto
+      auth_length: auto
+      opcode: 0x00000009
+      status: 0x0000
+    body_description: list opcodes result [1,3,2]
+    auth: 
+      type: direct
+      app_name: jimbob
```

Comment:
[Responses don't have authentication](https://parallaxsecond.github.io/parsec-book/parsec_client/wire_protocol.html#authentication) (or, if they did, it'd probably have to be an authenticator for the service) - but I think it makes sense to keep this open and supported since it's part of the wire protocol.

Reply:
fair point - may add a comment to say that, this isn't normal, but the test spec allows you to create whatever responses you want, valid or not.

Reply:
added comment abd changed enum on sample spec

Reply:
I disagree hugues - I was using the test data to check the error conditions for parsing on the golang client - negative testing is as important as positive.

We _can_ actually create any arbitrary data at the end of the header on request or response - we just need an appropriate generator function - they're designed to return binary data, though the examples so far give well formed protocol buffer data. 

We could create alternative authentication type of 'generate' which could take a generator function name for more complex authenticators.

Reply:
have, however, removed the auth from the response as incorrect